### ISSUE TITLE
CentOS 7 host fails on ansible.builtin.uri when matrix_nginx_proxy_ssl_preset: modern

### ISSUE THREAD

**Issue by pcfe** [spantaleev/matrix-docker-ansible-deploy#2364]
### Descripton

If I have `matrix_nginx_proxy_ssl_preset: modern` configured and I run `ansible-playbook setup.yml --tags=self-check` (on a Fedora 37 machine), then my CentOS 7 host fails with
```
Status code was -1 and not [200]: Request failed: <urlopen error [SSL: TLSV1_ALERT_PROTOCOL_VERSION] tlsv1 alert protocol version (_ssl.c:618)>
```

### Steps to reproduce

1. matrix server is a CentOS 7 machine
2. have `matrix_nginx_proxy_ssl_preset: modern` set
3. run `ansible-playbook setup.yml --tags=setup-all,ensure-matrix-users-created,start` to ensure the setting is applied
4. run `ansible-playbook setup.yml --tags=self-check`

### Actual result

```text
TASK [custom/matrix-nginx-proxy : Check .well-known on the matrix hostname] **************************************************
fatal: [matrix.pcfe.net]: FAILED! => changed=false 
  content: ''
  elapsed: 0
  msg: 'Status code was -1 and not [200]: Request failed: <urlopen error [SSL: TLSV1_ALERT_PROTOCOL_VERSION] tlsv1 alert protocol version (_ssl.c:618)>'
  redirected: false
  status: -1
  url: https://matrix.pcfe.net/.well-known/matrix/client
...ignoring

TASK [custom/matrix-nginx-proxy : Fail if .well-known not working on the matrix hostname] ************************************
fatal: [matrix.pcfe.net]: FAILED! => changed=false 
  msg: 'Failed checking that the well-known file for Client Discovery is configured at `matrix.pcfe.net` (checked endpoint: `https://matrix.pcfe.net/.well-known/matrix/client`). Is port 443 open in your firewall? Full error: {''status'': -1, ''url'': ''https://matrix.pcfe.net/.well-known/matrix/client'', ''changed'': False, ''elapsed'': 0, ''content'': '''', ''failed'': True, ''redirected'': False, ''msg'': ''Status code was -1 and not [200]: Request failed: <urlopen error [SSL: TLSV1_ALERT_PROTOCOL_VERSION] tlsv1 alert protocol version (_ssl.c:618)>''}'

PLAY RECAP *******************************************************************************************************************
matrix.pcfe.net            : ok=23   changed=0    unreachable=0    failed=1    skipped=11   rescued=0    ignored=1   
```

### Expected result

```
TASK [custom/matrix-nginx-proxy : Check .well-known on the matrix hostname]
```
does not error out just because my server still runs CentOS 7 and is set up to do only TLS 1.3

#### minimal reproducer

_matrix_servers_ points at my CentOS 7 matrix server. _t3600_ is a Fedora 37 box used for cross check. They are defined in separate inventories.

```yaml
- name: Test for uri and TLS 1.3 only
  hosts:
  - matrix_servers
  - t3600
  gather_facts: true
  vars:
    well_known_url_matrix: https://matrix.pcfe.net/.well-known/matrix/client
    some_other_url_of_mine: https://blog.pcfe.net/hugo/index.html
  tasks:
    - name: Check, with uri, a URL on a host doing TLS 1.2 and 1.3 but nothing else
      ansible.builtin.uri:
        url: "{{ some_other_url_of_mine }}"
        follow_redirects: none
        return_content: false
        validate_certs: true

    # this host does only TLS 1.3
    - name: Check, with uri, a URL a host only doing TLS 1.3
      ansible.builtin.uri:
        url: "{{ well_known_url_matrix }}"
        follow_redirects: none
        return_content: false
        validate_certs: true
      ignore_errors: true

    # try same URL with a different module
    - name: Check, with get_url, a URL a host only doing TLS 1.3
      ansible.builtin.get_url:
        url: "{{ well_known_url_matrix }}"
        dest: /tmp/foo.conf
        mode: '0440'
      ignore_errors: true
```

Running the above gives me the below, so this is more of a problem with ansible.builtin.uri than with matrix-docker-ansible-deploy itself.

```text
$ ansible-playbook --ask-vault-pass -i […]/ansible/inventories/matrix-docker-ansible-deploy-inventory banana.yml
Vault password: 
[WARNING]: Could not match supplied host pattern, ignoring: t3600

PLAY [Test for uri and TLS 1.3 only] *****************************************************************************************

TASK [Gathering Facts] *******************************************************************************************************
Sunday 25 December 2022  19:03:25 +0100 (0:00:00.045)       0:00:00.045 ******* 
ok: [matrix.pcfe.net]

TASK [Check, with uri, a URL on a host doing TLS 1.2 and 1.3 but nothing else] ***********************************************
Sunday 25 December 2022  19:03:28 +0100 (0:00:02.517)       0:00:02.562 ******* 
ok: [matrix.pcfe.net]

TASK [Check, with uri, a URL a host only doing TLS 1.3] **********************************************************************
Sunday 25 December 2022  19:03:29 +0100 (0:00:01.061)       0:00:03.624 ******* 
fatal: [matrix.pcfe.net]: FAILED! => {"changed": false, "elapsed": 0, "msg": "Status code was -1 and not [200]: Request failed: <urlopen error [SSL: TLSV1_ALERT_PROTOCOL_VERSION] tlsv1 alert protocol version (_ssl.c:618)>", "redirected": false, "status": -1, "url": "https://matrix.pcfe.net/.well-known/matrix/client"}
...ignoring

TASK [Check, with get_url, a URL a host only doing TLS 1.3] ******************************************************************
Sunday 25 December 2022  19:03:30 +0100 (0:00:00.800)       0:00:04.424 ******* 
fatal: [matrix.pcfe.net]: FAILED! => {"changed": false, "dest": "/tmp/foo.conf", "elapsed": 0, "msg": "Request failed: <urlopen error [SSL: TLSV1_ALERT_PROTOCOL_VERSION] tlsv1 alert protocol version (_ssl.c:618)>", "url": "https://matrix.pcfe.net/.well-known/matrix/client"}
...ignoring

PLAY RECAP *******************************************************************************************************************
matrix.pcfe.net            : ok=4    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=2   

Sunday 25 December 2022  19:03:31 +0100 (0:00:01.050)       0:00:05.474 ******* 
=============================================================================== 
Gathering Facts ------------------------------------------------------------------------------------------------------- 2.52s
Check, with uri, a URL on a host doing TLS 1.2 and 1.3 but nothing else ----------------------------------------------- 1.06s
Check, with get_url, a URL a host only doing TLS 1.3 ------------------------------------------------------------------ 1.05s
Check, with uri, a URL a host only doing TLS 1.3 ---------------------------------------------------------------------- 0.80s
```

Running that same minimal reproducer against a non-el7 host (tested with a Fedora 37 host) works just fine.

```text
$ ansible-playbook banana.yml
[WARNING]: Could not match supplied host pattern, ignoring: matrix_servers

PLAY [Test for uri and TLS 1.3 only] *****************************************************************************************

TASK [Gathering Facts] *******************************************************************************************************
Sunday 25 December 2022  19:08:02 +0100 (0:00:00.017)       0:00:00.017 ******* 
ok: [t3600]

TASK [Check, with uri, a URL on a host doing TLS 1.2 and 1.3 but nothing else] ***********************************************
Sunday 25 December 2022  19:08:06 +0100 (0:00:04.130)       0:00:04.147 ******* 
ok: [t3600]

TASK [Check, with uri, a URL a host only doing TLS 1.3] **********************************************************************
Sunday 25 December 2022  19:08:07 +0100 (0:00:01.531)       0:00:05.679 ******* 
ok: [t3600]

TASK [Check, with get_url, a URL a host only doing TLS 1.3] ******************************************************************
Sunday 25 December 2022  19:08:09 +0100 (0:00:01.363)       0:00:07.043 ******* 
changed: [t3600]

PLAY RECAP *******************************************************************************************************************
t3600                      : ok=4    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0   

Sunday 25 December 2022  19:08:10 +0100 (0:00:01.607)       0:00:08.650 ******* 
=============================================================================== 
Gathering Facts ------------------------------------------------------------------------------------------------------- 4.13s
Check, with get_url, a URL a host only doing TLS 1.3 ------------------------------------------------------------------ 1.61s
Check, with uri, a URL on a host doing TLS 1.2 and 1.3 but nothing else ----------------------------------------------- 1.53s
Check, with uri, a URL a host only doing TLS 1.3 ---------------------------------------------------------------------- 1.36s
```

I was looking unsuccessfully for a way to make ansible.builtin.uri work with that old el7 host, no luck. I'm kinda hoping you know of a way.

Maybe primitively using command to grab the file with e.g. `curl --tlsv1.3`  if the matrix server is CentOS 7 and `matrix_nginx_proxy_ssl_preset == "modern"` is an option but that also feels suboptimal, but at least it would allow the checks from `roles/custom/matrix-nginx-proxy/tasks/self_check_well_known_file.yml` to actually do their job.

I'm kinda hoping you know of a way to handle el7 hosts more elegantly for grabbing the well-known info. :-)

Obviously, if I dumbly disable that whole check;

```diff
pcfe@t3600 matrix-docker-ansible-deploy (pcfe-skip-self_check_well_known-for-el7-targets) $ git show HEAD
commit e8774553562ef2d94e8b58339c563692d05fe305 (HEAD -> pcfe-skip-self_check_well_known-for-el7-targets)
Author: Patrick C. F. Ernzer <pcfe@pcfe.net>
Date:   Sun Dec 25 17:04:04 2022 +0100

    no ansible.builtin.uri usage on el7 if modern ssl
    
    both ansible.builtin.uri and ansible.builtin.get_url fail when the
    host is el7 and matrix_nginx_proxy_ssl_preset is set to modern
    
    oddly enough, curl --tlsv1.3 manages to get the well_known_url_matrix content
    while a plain curl also fails ¯\_(ツ)_/¯

diff --git a/roles/custom/matrix-nginx-proxy/tasks/main.yml b/roles/custom/matrix-nginx-proxy/tasks/main.yml
index 39cec7e5..4518ff4d 100644
--- a/roles/custom/matrix-nginx-proxy/tasks/main.yml
+++ b/roles/custom/matrix-nginx-proxy/tasks/main.yml
@@ -33,6 +33,7 @@
     - ansible.builtin.include_tasks: "{{ role_path }}/tasks/self_check_well_known.yml"
   tags:
     - self-check
+  when: not (ansible_facts['os_family'] == "RedHat" and ansible_facts['distribution_major_version'] == "7" and matrix_nginx_proxy_ssl_preset == "modern")
 
 - name: Mark matrix-nginx-proxy role as executed
   ansible.builtin.set_fact:
```

It will stop the self-check from erroring out, but also skip the sensible checks from happening.

#### OS release info of host

```bash
[ansible@matrix ~]$ cat /etc/os-release 
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```

#### Testing with curl on the host

On the CentOS 7 host, I can curl the uri but only if I nail down to TLS 1.3

```
[ansible@matrix ~]$ curl -v https://matrix.pcfe.net/.well-known/matrix/client                   
* About to connect() to matrix.pcfe.net port 443 (#0)
*   Trying 138.201.144.34...
* Connected to matrix.pcfe.net (138.201.144.34) port 443 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
*   CAfile: /etc/pki/tls/certs/ca-bundle.crt
  CApath: none
* NSS error -12190 (SSL_ERROR_PROTOCOL_VERSION_ALERT)
* Peer reports incompatible or unsupported protocol version.
* Closing connection 0
curl: (35) Peer reports incompatible or unsupported protocol version.
```

```bash
[ansible@matrix ~]$ curl --tlsv1.3 -v https://matrix.pcfe.net/.well-known/matrix/client         
* About to connect() to matrix.pcfe.net port 443 (#0)
*   Trying 138.201.144.34...
* Connected to matrix.pcfe.net (138.201.144.34) port 443 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
*   CAfile: /etc/pki/tls/certs/ca-bundle.crt
  CApath: none
* SSL connection using TLS_AES_128_GCM_SHA256
* Server certificate:
*       subject: CN=matrix.pcfe.net
*       start date: Nov 27 02:02:52 2022 GMT
*       expire date: Feb 25 02:02:51 2023 GMT
*       common name: matrix.pcfe.net
*       issuer: CN=R3,O=Let's Encrypt,C=US
> GET /.well-known/matrix/client HTTP/1.1
> User-Agent: curl/7.29.0
> Host: matrix.pcfe.net
> Accept: */*
> 
< HTTP/1.1 200 OK
< Server: nginx
< Date: Sun, 25 Dec 2022 17:47:37 GMT
< Content-Type: application/json
< Content-Length: 434
< Last-Modified: Mon, 21 Mar 2022 23:02:28 GMT
< Connection: keep-alive
< ETag: "62390404-1b2"
< Expires: Sun, 25 Dec 2022 21:47:37 GMT
< Cache-Control: max-age=14400
< Access-Control-Allow-Origin: *
< Accept-Ranges: bytes
< 
{
    "im.vector.riot.e2ee": {
        "default": true
    },
    "im.vector.riot.jitsi": {
        "preferredDomain": "jitsi.pcfe.net"
    },
    "io.element.e2ee": {
        "default": true,
        "secure_backup_required": true,
        "secure_backup_setup_methods": []
    },
    "io.element.jitsi": {
        "preferredDomain": "jitsi.pcfe.net"
    },
    "m.homeserver": {
        "base_url": "https://matrix.pcfe.net"
    }
* Connection #0 to host matrix.pcfe.net left intact
```

#### Why still on CentOS 7

Basically I'm just waiting for #520 to happen, once that is there, I'll replace the affected _CentOS 7_ machine with a _CentOS Stream 9_ machine

#### SSL Labs Reports for the two tested sites of mine

I think that TLS 1.3 only is the trigger for this issue, here's the Qualsys results if you would like to crosscheck it's not something else I overlooked

- <https://www.ssllabs.com/ssltest/analyze.html?d=matrix.pcfe.net>
- <https://www.ssllabs.com/ssltest/analyze.html?d=blog.pcfe.net>

**Comment by spantaleev** [comment]
You'd better replace CentOS 7 sooner or stop running Ansible on the CentOS server. CentOS is dead and CentOS 7 is damn ancient and you can see that in its crappy outdated SSL support.

#520 is unlikely to happen any time soon.

Kill CentOS 7 with fire now ;) You know IBM did

**Comment by pcfe** [comment]
> […] stop running Ansible on the CentOS server.[…]

I do not, the task is executed on the host, not on the ansible control  node.
I've though about `delegat_to` this one task to locahost but that feels overly hacky to deal with el7 and `ansible.builtin.uri` 

>[CentOS 7] outdated SSL support.

Fair enough. Feel free to CLOSE WONTFIX

> #520 is unlikely to happen any time soon.

Shame, but thanks for the update.
